feat(menu): Added ability to show the menu overlay around the menu trigger

trshafer opened the pull request:
Currently the menu is fixed to show over the menu trigger.

This adds a (rather blunt) boolean to decide if the menu should be shown above or below the trigger. I thought about changing the possibilities to `overlay-position: default | above | below | left | right;` But that felt like a bit too much flexibility. I think some version of this feature is really important for those designing their own menu. Let me know what you think.

Thanks,
Thomas

Here is the menu below the trigger:
![Menu Below](https://cloud.githubusercontent.com/assets/135461/20110876/fc6db274-a59a-11e6-83a8-0dd0de59193c.png)

Here is the menu above the trigger:
![Menu Above](https://cloud.githubusercontent.com/assets/135461/20110875/fa3c06f4-a59a-11e6-95b4-04731f153bc6.png)

jelbourn commented on src/lib/menu/README.md:
```diff
@@ -114,8 +114,9 @@ Output:
-using the `x-position` (`before | after`) and `y-position` (`above | below`) attributes.  
+By default, the menu will display after and below its trigger.  You can change this display position
+using the `x-position` (`before | after`) and `y-position` (`above | below`) attributes. The menu
+can be positioned over the menu button or outside using `overlay-trigger` (`true | false`).
```
I'm not a fan of this name ("overlay-trigger") 

@kara thoughts?

kara replied:
How about "overlap-trigger"?  "Overlay-trigger" might sound like the trigger itself is an overlay.

jelbourn commented:
R: @kara

kara commented on src/lib/menu/menu-directive.ts:
```diff
@@ -46,14 +46,18 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
   constructor(@Attribute('x-position') posX: MenuPositionX,
-              @Attribute('y-position') posY: MenuPositionY) {
+              @Attribute('y-position') posY: MenuPositionY,
+              @Attribute('overlay-trigger') trigger = true) {
```
Have you considered using the `public` keyword instead?  That way you don't have to explicitly set it again on line 60.

```
@Attribute('overlap-trigger') public overlapTrigger = true
```

trshafer replied:
Thanks for the suggestion.

kara commented on src/lib/menu/README.md:
```diff
@@ -148,7 +149,8 @@ also adds `aria-hasPopup="true"` to the trigger element.
 | `x-position` | `before | after` | The horizontal position of the menu in relation to the trigger. Defaults to `after`. | 
 | `y-position` | `above | below` | The vertical position of the menu in relation to the trigger. Defaults to `below`. |
- 
+| `overlay-trigger`    | `true | false` | Whether to have the menu show on top of the menu trigger or outside. Defaults to `true`. |
```
Can we rename to `overlapTrigger`?

trshafer replied:
Done.

kara commented on src/lib/menu/menu-trigger.ts:
```diff
@@ -197,12 +197,15 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
+    let originY = overlayY;
+    if (this.menu.overlayTrigger) {
+      originY = overlayY === 'top' ? 'bottom' : 'top';
+    }
```
If you want to overlap the trigger, shouldn't the originY and the overlayY be the same? e.g. the top of the trigger should connect to the top of the overlay or the bottom of the trigger should connect to the bottom of the overlay.  Let me know if I'm missing something.

```ts
let originY: VerticalConnectionPos;
if (this.menu.overlapTrigger) {
   originY = overlayY;
} else {
   originY = overlayY === 'top' ? 'bottom' : 'top';
}
```

trshafer replied:
One would think so. However I believe the current implementation produces the desired effect. Are you able to clone the pr and run the demo-app and verify? I have done so and I believe the functionality is accurate.

trshafer commented:
Ah, looks like there's some test bleedover. If I just focus on the [fallback positions describe block](https://github.com/angular/material2/blob/master/src/lib/menu/menu.spec.ts#L137) using fdescribe, those tests will fail. If I focus on the [whole menu test](https://github.com/angular/material2/blob/master/src/lib/menu/menu.spec.ts#L22) then the fallback positions test passes.

trshafer commented:
I have logged the issue: #2530.

trshafer commented:
Hey @kara, I'm happy to write tests for the updated change. However since the tests are currently failing I don't feel confident in updating the tests or making new ones without potentially causing issues. Are you able to update the tests? I could attempt to fix the tests, however since I don't have the background knowledge in those tests, I thought it would be better to defer to the original author. Looks like they were introduced here: https://github.com/angular/material2/commit/7572e345dca18e139440394c6c8671ecabfaf1a9

kara commented:
@trshafer We won't be able to merge this PR without tests. It's standard practice for this repo to provide tests with your code. Go ahead and give it a shot. I can take a look at your other issue in the meantime.

trshafer commented:
Totally. Happy to write tests. I should have red, green, refactored my way to this PR. Starting from a red test is harder for me to confidently ensure the test and change I have written is correct. It would be great if the menu test suite was green before I start.

I've also rebased in the recent commit.

kara commented:
Yep, that makes sense. I'll ping you once I've looked at the the existing tests.

trshafer commented:
Hey @kara, thanks for submitting the test fixes. I have written two tests for the overlap change.

I made an OverlapSubject class to reduce the testing duplication. Currently its visibility is limited to the new test describe block, however it might be helpful for the other tests as well.

trshafer commented on src/lib/menu/menu.spec.ts:
```diff
@@ -256,6 +256,66 @@ describe('MdMenu', () => {
   });
 
+  describe('not overlapping', () => {
+    class OverlapSubject<T extends TestableMenu> {
```
Done.

trshafer commented on src/lib/menu/menu.spec.ts:
```diff
@@ -256,6 +256,66 @@ describe('MdMenu', () => {
+      }
+
+      updateTriggerStyle(style: any) {
+        return Object.assign(this.trigger.style, style);
```
Done.

trshafer commented on src/lib/menu/menu.spec.ts:
```diff
@@ -256,6 +256,66 @@ describe('MdMenu', () => {
+      // Since the menu is below the trigger, the overlay top should be the trigger bottom.
+      expect(Math.round(subject.overlayRect.top))
+          .toBe(Math.round(subject.triggerRect.bottom),
+              `Expected menu to open in "above" position if "below" position wouldn't fit.`);
```
Done.

trshafer commented on src/lib/menu/menu.spec.ts:
```diff
@@ -256,6 +256,66 @@ describe('MdMenu', () => {
+          .toBe(Math.round(subject.triggerRect.top),
+              `Expected menu to open in "above" position if "below" position wouldn't fit.`);
+    });
+  });
```
Done.

trshafer commented on src/lib/menu/menu.spec.ts:
```diff
@@ -311,6 +371,22 @@ class PositionedMenu {
+  `
+})
+class NonOverlapMenu implements TestableMenu {
+  @ViewChild(MdMenuTrigger) trigger: MdMenuTrigger;
```
It's being used to open the menu: https://github.com/trshafer/material2/blob/5373c66e8cd9d0a06485158434d0afbc5233245a/src/lib/menu/menu.spec.ts#L271

kara replied:
I see. Totally missed that!

trshafer commented on src/lib/menu/menu-trigger.ts:
```diff
@@ -230,21 +230,29 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
+
+    if (this.menu.overlapTrigger) {
+      originY = overlayY === 'top' ? 'bottom' : 'top';
+      fallbackOriginY = fallbackOverlayY === 'top' ? 'bottom' : 'top';
```
You are absolutely correct. The test of an explicit overlapTrigger = true is a good suggestion and exposes this in a test. Thanks for pushing back. Fixed.

trshafer commented on src/lib/menu/README.md:
```diff
@@ -115,7 +115,8 @@ Output:
 By default, the menu will display after and below its trigger.  You can change this display position
-using the `x-position` (`before | after`) and `y-position` (`above | below`) attributes.
+using the `x-position` (`before | after`) and `y-position` (`above | below`) attributes. The menu
+can be positioned over the menu button or outside using `overlap-trigger` (`true | false`).
```
Updated. Let me know if there is more documentation I should add.

trshafer commented on src/lib/menu/menu-directive.ts:
```diff
@@ -54,7 +54,8 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
+              @Attribute('y-position') posY: MenuPositionY,
+              @Attribute('overlap-trigger') public overlapTrigger = true) {
     if (posX) { this._setPositionX(posX); }
     if (posY) { this._setPositionY(posY); }
```
Good catch. I have updated the menu positioning classes and added a test.

trshafer commented:
Hello @kara, thank you for the thorough review. I'm pretty sure I have responded to all of your comments. Let me know what else I can do to help move this across the finish line.

mmalerba commented on src/lib/menu/menu-directive.ts:
```diff
@@ -52,6 +52,7 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
 
   @ViewChild(TemplateRef) templateRef: TemplateRef<any>;
   @ContentChildren(MdMenuItem) items: QueryList<MdMenuItem>;
+  @Input('overlap-trigger') overlapTrigger = true;
```
Don't need to rename to dash-case, we're trying to switch everything to camelCase

trshafer replied:
so should the api look like `<md-menu [overlapTrigger]="false">`?

trshafer commented:
Switched `overlap-trigger` to `overlapTrigger`.

mmalerba commented on src/lib/menu/menu-directive.ts:
```diff
@@ -52,6 +52,7 @@ export class MdMenu implements AfterContentInit, MdMenuPanel, OnDestroy {
 
   @ViewChild(TemplateRef) templateRef: TemplateRef<any>;
   @ContentChildren(MdMenuItem) items: QueryList<MdMenuItem>;
+  @Input('overlapTrigger') overlapTrigger = true;
```
tiny nit: if the variable and the html attr have the same name you can omit the alias:
`@Input() overlapTrigger = true;

trshafer replied:
Done.
